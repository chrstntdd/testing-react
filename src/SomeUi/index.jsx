import React from 'react';
import PropTypes from 'prop-types';

export class SomeUi extends React.Component {
  static propTypes = {
    propFuncReq: PropTypes.func.isRequired,
    propFuncOpt: PropTypes.func,
    propA: PropTypes.shape({
      key1: PropTypes.bool.isRequired,
      key2: PropTypes.oneOf(['variant1', 'variant2', 'variant3']).isRequired
    }).isRequired,
    propB: PropTypes.instanceOf(Map).isRequired,
    propC: PropTypes.bool.isRequired
  };

  static defaultProps = {
    propFuncOpt: () => {}
  };

  state = { inputValue: '', hidden: true };

  componentDidUpdate(prevProps) {
    // if the prop has changed between updates, do side effect-y things
    if (this.props.propC !== prevProps.propC) {
      this.props.propFuncReq();
    }
  }

  componentWillUnmount() {
    // cleanup function
    this.props.propFuncOpt();
  }

  handleInputChange = e => {
    const inputValue = e.target.value;

    this.setState({ inputValue: inputValue.trim() });
  };

  toggleVisibility = () => {
    this.setState(prevState => ({ hidden: !prevState.hidden }));
  };

  render() {
    return (
      <div>
        <input
          className="test-input"
          placeholder="Enter some text whydontcha"
          onChange={this.handleInputChange}
        />
        <br />
        {this.state.inputValue && <p>this.state.inputValue</p>}
        <br />
        <button onClick={this.toggleVisibility}>{this.state.hidden ? 'show' : 'hide'}</button>
        <br />
        {!this.state.hidden && <p>hidden content</p>}
      </div>
    );
  }
}

export default SomeUi;
