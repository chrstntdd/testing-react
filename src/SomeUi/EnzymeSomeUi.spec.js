import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import SomeUi from '.';

configure({ adapter: new Adapter() });

describe('Enzyme testing', () => {
  let props;

  beforeEach(() => {
    props = {
      propFuncReq: jest.fn(),
      propA: { key1: true, key2: 'variant1' },
      propB: new Map([['a', 'b'], [1, 2]]),
      propC: false
    };
  });

  it('should work', () => {
    const wrapper = shallow(<SomeUi {...props} />);

    expect(wrapper).toBeDefined();
  });

  describe('toggleVisibility method', () => {
    const render = SomeUi.prototype.render;

    beforeAll(() => {
      SomeUi.prototype.render = jest.fn();
    });

    afterAll(() => {
      SomeUi.prototype.render = render;
    });

    it('should toggle the visibility', () => {
      const wrapper = shallow(<SomeUi {...props} />);

      wrapper.instance().toggleVisibility();

      expect(wrapper.state().hidden).toBe(false);
    });
  });
});
