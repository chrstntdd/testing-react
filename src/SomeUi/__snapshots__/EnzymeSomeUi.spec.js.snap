// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enzyme testing render method should render some content 1`] = `
<div>
  <input
    className="test-input"
    onChange={[Function]}
    placeholder="Enter some text whydontcha"
  />
  <br />
  <br />
  <button
    onClick={[Function]}
  >
    show
  </button>
  <br />
</div>
`;
